mmv1/third_party/terraform/services/resourcemanager/ephemeral_google_service_account_token.go

```diff
@@ -0,0 +1,275 @@
+
+	ScopesSetValue, diags := data.Scopes.ToSetValue(ctx)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	var delegates []string
+	if !data.Delegates.IsNull() {
+		DelegatesSetValue, diags := data.Delegates.ToSetValue(ctx)
+		resp.Diagnostics.Append(diags...)
+		if resp.Diagnostics.HasError() {
+			return
+		}
+		delegates = StringSet(DelegatesSetValue)
+	}
+
+	tokenRequest := &iamcredentials.GenerateAccessTokenRequest{
+		Lifetime:  data.Lifetime.ValueString(),
+		Delegates: delegates,
+		Scope:     tpgresource.CanonicalizeServiceScopes(StringSet(ScopesSetValue)),
+	}
+
+	at, err := service.Projects.ServiceAccounts.GenerateAccessToken(name, tokenRequest).Do()
+	if err != nil {
+		resp.Diagnostics.AddError(
+			"Error generating access token",
+			fmt.Sprintf("Error generating access token: %s", err),
+		)
+		return
+	}
+
+	data.AccessToken = types.StringValue(at.AccessToken)
+	resp.Diagnostics.Append(resp.Result.Set(ctx, data)...)
+}
+
+func StringSet(d basetypes.SetValue) []string {
+
+	StringSlice := make([]string, 0)
+	for _, v := range d.Elements() {
+		StringSlice = append(StringSlice, v.(basetypes.StringValue).ValueString())
+	}
+	return StringSlice
+}
+
+// Define the possible service account name patterns
+var serviceAccountNamePatterns = []string{
+	`^.+@.+\.iam\.gserviceaccount\.com$`,                     // Standard IAM service account
+	`^.+@developer\.gserviceaccount\.com$`,                   // Legacy developer service account
+	`^.+@appspot\.gserviceaccount\.com$`,                     // App Engine service account
+	`^.+@cloudservices\.gserviceaccount\.com$`,               // Google Cloud services service account
+	`^.+@cloudbuild\.gserviceaccount\.com$`,                  // Cloud Build service account
+	`^service-[0-9]+@.+-compute\.iam\.gserviceaccount\.com$`, // Compute Engine service account
+}
+
+// Create a custom validator for service account names
+type serviceAccountNameValidator struct{}
+
+func (v serviceAccountNameValidator) Description(ctx context.Context) string {
+	return "value must be a valid service account email address"
+}
+
+func (v serviceAccountNameValidator) MarkdownDescription(ctx context.Context) string {
+	return v.Description(ctx)
+}
+
+func (v serviceAccountNameValidator) ValidateString(ctx context.Context, req validator.StringRequest, resp *validator.StringResponse) {
+	if req.ConfigValue.IsNull() || req.ConfigValue.IsUnknown() {
+		return
+	}
+
+	value := req.ConfigValue.ValueString()
+	valid := false
+	for _, pattern := range serviceAccountNamePatterns {
+		if matched, _ := regexp.MatchString(pattern, value); matched {
+			valid = true
+			break
+		}
+	}
+
+	if !valid {
+		resp.Diagnostics.AddAttributeError(
+			req.Path,
+			"Invalid Service Account Name",
+			"Service account name must match one of the expected patterns for Google service accounts",
+		)
+	}
+}
+
+// Create a custom validator for duration
+type durationValidator struct {
```

Comment:
Could you please add some automated tests for this new validator?

Reply:
refer to https://github.com/GoogleCloudPlatform/magic-modules/pull/12140#discussion_r1833467151